testhelper/tests.go

```diff
@@ -231,121 +267,189 @@ func (options *TestOptions) RunTestUpgrade() (*terraform.PlanStruct, error) {
+		// Backup the original value of DisableTempWorkingDir
+		tempDirCreationBackup := options.DisableTempWorkingDir
+
+		// Temporarily disable the creation of a temporary directory
+		options.DisableTempWorkingDir = true
+
+		// Defer a function to restore the original value
+		defer func() {
+			options.DisableTempWorkingDir = tempDirCreationBackup
+		}()
+
 		// Setup the test including a TEMP directory to run in
 		options.testSetup()
 
-		// Create a temporary directory for the state file
-		tempDir, err := os.MkdirTemp("", fmt.Sprintf("terraform-%s", options.Prefix))
+		// Create a temporary directory for the PR code
+		prTempDir, err := os.MkdirTemp("", fmt.Sprintf("terraform-pr-%s", options.Prefix))
 		if err != nil {
 			logger.Log(options.Testing, err)
+		} else {
+			logger.Log(options.Testing, "TEMP PR DIR CREATED: ", prTempDir)
 		}
-		defer os.RemoveAll(tempDir) // clean up
-
-		// from here on we will operate in the temp directory
-		gitRoot, _ := common.GitRootPath(options.TerraformDir)
-		gitRepo, _ := git.PlainOpen(gitRoot)
+		defer os.RemoveAll(prTempDir) // clean up
 
-		// maintain a reference of current checkout, which might be a detatched PR merge, will be used to switch back later
-		prRef, _ := gitRepo.Head()
-		logger.Log(options.Testing, "Current Branch [Name - Hash]:", prRef.Name(), "-", prRef.Hash())
-
-		// fetch to ensure all branches are present
-		remote, err := gitRepo.Remote("origin")
+		// Create a temporary directory for the base branch
+		baseTempDir, err := os.MkdirTemp("", fmt.Sprintf("terraform-base-%s", options.Prefix))
 		if err != nil {
 			logger.Log(options.Testing, err)
+		} else {
+			logger.Log(options.Testing, "TEMP UPGRADE BASE DIR CREATED: ", baseTempDir)
 		}
+		defer os.RemoveAll(baseTempDir) // clean up
 
-		opts := &git.FetchOptions{
-			RefSpecs: []config.RefSpec{"refs/*:refs/*"},
+		// Copy the current code (from PR branch) to the PR temp directory with the filter
+		errCopy := common.CopyDirectory(gitRoot, prTempDir, func(path string) bool {
+			if files.PathContainsTerraformStateOrVars(path) {
+				return false
+			}
+			// Add a filter to ignore directories named "temp"
+			if strings.Contains(strings.ToLower(path), "/temp/") || strings.HasSuffix(strings.ToLower(path), "/temp") {
+				return false
+			}
+			return true
+		})
+		if errCopy != nil {
+			// Tear down the test
+			options.testTearDown()
+			return nil, fmt.Errorf("failed to copy PR directory to temp: %v", errCopy)
+		} else {
+			logger.Log(options.Testing, "Copied current code to PR branch dir:", prTempDir)
 		}
 
-		if err := remote.Fetch(opts); err != nil {
-			logger.Log(options.Testing, err)
+		baseRepo, baseBranch := common.GetBaseRepoAndBranch(options.BaseTerraformRepo, options.BaseTerraformBranch)
+		if baseBranch == "" || baseRepo == "" {
+			return nil, fmt.Errorf("failed to get default repo and branch: %s %s", baseRepo, baseBranch)
```

Comment:
Some error conditions in the upgrade function call `testTearDown()` and some do not, should all error returns have a `testTearDown()` as well?